applications/FluidDynamicsApplication/custom_processes/shock_capturing_process.cpp

```diff
@@ -0,0 +1,442 @@
+            // Perform the elemental loop
+            block_for_each(mrModelPart.Elements(), tls_container_2D3N, aux_function_2D3N);
+        } else if (geometry_type == GeometryData::KratosGeometryType::Kratos_Tetrahedra3D4) {
+            // Set auxiliary TLS container and elemental function
+            ShockCapturingTLSType3D4N tls_container_3D4N = SetTLSContainer3D4N();
+            auto aux_function_3D4N = [&, this] (Element &rElement, ShockCapturingTLSType3D4N &rShockCapturingTLS) {this->CalculatePhysicsBasedShockCapturingElementContribution<3,4>(rElement, rShockCapturingTLS);};
+            // Perform the elemental loop
+            block_for_each(mrModelPart.Elements(), tls_container_3D4N, aux_function_3D4N);
+        } else {
+            KRATOS_ERROR << "Asking for a non-supported geometry. Physics-based shock capturing only supports \'Triangle2D3\' and \'Tetrahedra3D4\' geometries.";
+        }
+
+        // Nodal smoothing of the shock capturing magnitudes
+        // Note that to avoid calculating the NODAL_AREA we took from the first DOF of the lumped mass vector
+        IndexPartition<>(n_nodes).for_each([&](const ModelPart::SizeType iNode) {
+            auto it_node = mrModelPart.NodesBegin() + iNode;
+            const double nodal_area = it_node->GetValue(NODAL_AREA);
+            it_node->GetValue(ARTIFICIAL_CONDUCTIVITY) /= nodal_area;
+            it_node->GetValue(ARTIFICIAL_BULK_VISCOSITY) /= nodal_area;
+            it_node->GetValue(ARTIFICIAL_DYNAMIC_VISCOSITY) /= nodal_area;
+        });
+    }
+
+    double ShockCapturingProcess::LimitingFunction(
+        const double s,
+        const double s_0,
+        const double s_max,
+        const double s_min)
+    {
+        const double aux_1 = std::max(s - s_0, s_min);
+        const double aux_2 = std::min(aux_1 - s_max, s_min);
+        return aux_2 + s_max;
+    }
+
+    double ShockCapturingProcess::SmoothedLimitingFunction(
+        const double s,
+        const double s_0,
+        const double s_max)
+    {
+        const double aux_1 = SmoothedMaxFunction(s - s_0);
+        const double aux_2 = SmoothedMinFunction(aux_1 - s_max);
+        return aux_2 + s_max;
+    }
+
+    // Smooth approximation of the max(s,0) function
+    double ShockCapturingProcess::SmoothedMaxFunction(const double s)
+    {
+        const double b = 100;
+        const double l_max = (s / Globals::Pi) * std::atan(b * s) + 0.5 * s - (1.0 / Globals::Pi) * std::atan(b) + 0.5;
+        return l_max;
+    }
+
+    // Smooth approximation of the min(s,0) function
+    double ShockCapturingProcess::SmoothedMinFunction(const double s)
+    {
+        return s - SmoothedMaxFunction(s);
+    }
+
+
+    double ShockCapturingProcess::CalculateProjectedInverseMetricElementSize(
+        const Matrix& rInverseMetricTensor,
```

Comment:
no hope to use bounded_matrix?

Reply:
This would require changing the `MathUtils` `InvertMatrix` method.

Reply:
there is a templated version of that

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/utilities/math_utils.h#L425

Reply:
i will approve anyway, but it would be good to eventually change it

Reply:
I think this is the one I tried to use but it gave me a compiler error (let me just try again before merging).

Reply:
Fixed. I can't remember what happened...